gammapy/spectrum/energy.py

```diff
@@ -0,0 +1,293 @@
+        The unit of the value specified for the energy.  This may be
+        any string that `~astropy.units.Unit` understands, but it is
+        better to give an actual unit object.
+
+    dtype : `~numpy.dtype`, optional
+        See `~astropy.units.Quantity`.
+
+    copy : bool, optional
+        See `~astropy.units.Quantity`.
+
+    """
+
+    def __new__(cls, energy, unit=None, dtype=None, copy=True):
+
+        # Techniques to subclass Quantity taken from astropy.coordinates.Angle
+        # see: http://docs.scipy.org/doc/numpy/user/basics.subclassing.html
+        if unit is not None:
+            unit = u.Unit(unit)
+            if not unit.is_equivalent(u.eV):
+                raise ValueError("Requested unit {0} is not an"
+                                 " energy unit".format(unit))
+
+        if isinstance(energy, u.Quantity):
+            # also True if energy is of type Energy
+            if unit is not None:
+                energy = energy.to(unit).value
+            else:
+                unit = energy.unit
+                if not unit.is_equivalent(u.eV):
+                    raise ValueError("Given quantity {0} is not an"
+                                     " energy".format(energy))
+                energy = energy.value
+
+        else:
+            if unit is None:
+                raise ValueError("No unit given")
+
+        self = super(Energy, cls).__new__(cls, energy, unit,
+                                          dtype=dtype, copy=copy)
+
+        # Interesting once  energy bounds are stored
+        self._nbins = self.size
+
+        return self
+
+    def __array_finalize__(self, obj):
+        if obj is None:
+            return
+        self._nbins = self.size
+        self._unit = getattr(obj, '_unit', None)
+
+    def __quantity_subclass__(self, unit):
+        unit = u.Unit(unit)
+        if unit.is_equivalent(u.eV):
+            return Energy, True
+        else:
+            return super(Energy, self).__quantity_subclass__(unit)[0], False
+
+    @property
+    def nbins(self):
+        """
+        The number of bins
+        """
+        return self._nbins
+
+    @staticmethod
+    def equal_log_spacing(emin, emax, nbins, unit=None):
+        """Create Energy with equal log-spacing.
+
+        if no unit is given, it will be taken from emax
+
+        Parameters
+        ----------
+        emin : `~astropy.units.Quantity`, float
+            Lowest energy bin
+        emax : `~astropy.units.Quantity`, float
+            Highest energy bin
+        bins : int
+            Number of bins
+        unit : `~astropy.units.UnitBase`, str
+            Energy unit
+
+        Returns
+        -------
+        Energy
+        """
+
+        if unit is None:
+            unit = emax.unit
+
+        emin = Quantity(emin, unit)
+        emax = Quantity(emax, unit)
+
+        x_min, x_max = np.log10([emin.value, emax.value])
+        energy = np.logspace(x_min, x_max, nbins)
+
+        return Energy(energy, unit)
```

Comment:
Slight speed-up by specifying `copy=False` here (OK, since you know you just created the array).